Convert `run-make/coverage-reports` tests to use a custom compiletest mode

Zalathar opened the pull request:
I was frustrated by the fact that most of the coverage tests are glued together with makefiles and shell scripts, so I tried my hand at converting most of them over to a newly-implemented `run-coverage` mode/suite in compiletest.

This ~~*mostly*~~ resolves #85009, ~~though I've left a small number of the existing tests as-is because they would require more work to fix/support~~.

---

I had time to go back and add support for the more troublesome tests that I had initially skipped over, so this PR now manages to completely get rid of `run-make/coverage-reports`.

---

The patches are arranged as follows:

- Declare the new mode/suite in bootstrap
- Small changes to compiletest that will be used by the new mode
- Implement the new mode in compiletest
- Migrate most of the tests over
- Add more code to bootstrap and compiletest to support the remaining tests
- Migrate the remaining tests (with some temporary hacks to avoid re-blessing them)
- Remove the temporary hacks and re-bless the migrated tests
- Remove the unused remnants of `run-make/coverage-reports`

Zalathar commented:
The part I’m least confident about is the implied `needs-profiler-support` check for each test, since the existing needs-checking code doesn't really anticipate that sort of use-case.

Zalathar commented:
@rustbot +A-code-coverage

Zalathar commented:
Something to note is that the existing makefile has three directives:

- `needs-profiler-support`
- `ignore-windows-gnu`
- `ignore-cross-compile`

I've translated `ignore-cross-compile` into a `host_test!` declaration in bootstrap, and I added an implicit profiler-support check as mentioned above.

I haven't translated the `ignore-windows-gnu` directive into anything, since I don't quite understand what (if anything) needs to be done there.

Zalathar commented:
@rustbot label +A-code-coverage

wesleywiser commented:
> The part I’m least confident about is the implied needs-profiler-support check for each test, since the existing needs-checking code doesn't really anticipate that sort of use-case.

This part looks ok to me but if @jyn514 wants to check as well, that would be awesome!

> ignore-windows-gnu

If I recall correctly, code coverage isn't supported on windows-gnu which is why the tests ignore it.

Zalathar commented on src/tools/compiletest/src/header/needs.rs:
```diff
@@ -190,6 +190,7 @@ pub(super) struct CachedNeedsConditions {
     sanitizer_memtag: bool,
     sanitizer_shadow_call_stack: bool,
     sanitizer_safestack: bool,
+    pub(super) profiler_support: bool,
```
The `pub(super)` isn’t needed immediately, but it is needed later on when I implement the `run-coverage` mode, because the new code that checks for it is in the parent module.

(This is admittedly a bit of a hack, but in this case I figured a narrow hack was preferable to redesigning the needs-checker for one use-case.)

Zalathar commented on src/bootstrap/test.rs:
```diff
@@ -1592,7 +1594,9 @@ note: if you're sure you want to do this, please open an issue as to why. In the
             // Only pass correct values for these flags for the `run-make` suite as it
             // requires that a C++ compiler was configured which isn't always the case.
-            if !builder.config.dry_run() && matches!(suite, "run-make" | "run-make-fulldeps") {
+            if !builder.config.dry_run()
```
I was reluctant to touch this comment because it seems to already be somewhat out-of-date (the C++ compiler parts have been moved elsewhere), and I‘m not quite sure what to replace it with. But I suppose leaving it as-is isn't great either.

Zalathar commented:
Rebased and added some small tweaks based on feedback:

- Adjusted comments and control flow in the bootstrap code that supplies `--llvm-bin-dir`
- `profiler_support` is made `pub(super)` in the patch that actually needs it
- Added a comment to the `profiler_support` check

Zalathar commented:
In light of still needing to support `ignore-windows-gnu`, what I might do is add a way to inject “extra” directives into `iter_header`, so that they are processed as though they appeared directly in the file. That seems like the least intrusive way to support arbitrary mode-implied directives.

Zalathar commented:
OK, I’ve fundamentally changed how the mode-implied directives work.

`iter_header_extra` now takes a list of caller-supplied directives that should be processed as though they appeared on “line 0” of the file being inspected. There's a hard-coded check that adds three extra directives for `Mode::RunCoverage` tests, and does nothing for all other modes.

This is still a little bit gross, but I think it’s the best compromise for now.

Zalathar commented:
(A side-effect of the above is that it’s no longer necessary to move `profiler_support` into the needs cache struct, but I’ve left that change intact because I think it’s a reasonable little improvement in its own right.)

Zalathar commented:
I've now migrated *all* of the `run-make/coverage-reports` tests, so some code has been changed/added since the last review.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5239040642/jobs/9458510864) [(plain)](https://github.com/rust-lang/rust/commit/d96cc6e2c0bd68af0d6654ccf4e057e9872f8a91/checks/14174434352/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling compiletest v0.0.0 (/checkout/src/tools/compiletest)
error[E0308]: mismatched types
   --> src/tools/compiletest/src/runtest/tests.rs:114:47
    |
114 |     assert_eq!(TestCx::sort_coverage_subviews(&input), Ok(expected.to_owned()));
    |                ------------------------------ ^^^^^^ types differ in mutability
    |                arguments to this function are incorrect
    |
    = note: expected mutable reference `&mut Vec<&str>`
                       found reference `&&str`
                       found reference `&&str`
note: associated function defined here
   --> src/tools/compiletest/src/runtest.rs:685:8
    |
685 |     fn sort_coverage_subviews(coverage_lines: &mut Vec<&str>) -> Result<(), String> {

error[E0308]: mismatched types
   --> /checkout/library/core/src/macros/mod.rs:40:35
    |
    |
36  | macro_rules! assert_eq {
    | ---------------------- in this expansion of `assert_eq!`
...
40  |                 if !(*left_val == *right_val) {
    |                                   ^^^^^^^^^^ expected `Result<(), String>`, found `Result<String, _>`
   ::: src/tools/compiletest/src/runtest/tests.rs:114:5
    |
    |
114 |     assert_eq!(TestCx::sort_coverage_subviews(&input), Ok(expected.to_owned()));
    |
    = note: expected enum `std::result::Result<(), std::string::String>`
               found enum `std::result::Result<std::string::String, _>`

```

</details>

Zalathar commented:
Via https://github.com/rust-lang/rust/pull/112514#issuecomment-1586557025 it looks like my glob to find all the doctest executables doesn't work on Windows, because those directories also contain `.pdb` files that `llvm-cov show` rightly complains about.

I'll need to fix this by being smarter about detecting which files are executables.

Zalathar committed:
Add a custom `run-coverage` mode to compiletest

Zalathar committed:
Extract a common function for setting up environment vars

Zalathar committed:
Remember whether `failure-status` was explicitly specified

Currently a test without a `failure-status` directive is treated as having an
expected failure-status of 1, but `run-coverage` tests will want to treat those
tests as expecting success instead.

Zalathar committed:
Expand `run-coverage` to support the remaining `coverage-reports` tests

Zalathar committed:
Migrate the remaining `run-make/coverage-reports` tests over to `run-coverage`

To make it easier to verify that the output snapshots have been migrated
faithfully, this change adds some temporary helper code that lets us avoid
having to completely re-bless the existing snapshots.

A later change in this PR will then re-bless the tests and remove the temporary
helper code.

Zalathar committed:
Introduce `exec_compiled_test_general`

This will allow the `run-coverage` mode to easily set environment variable
`LLVM_PROFILE_FILE`, and to prevent the executable from being deleted after a
successful run.

Zalathar committed:
Move the `RUSTC_PROFILER_SUPPORT` check into `CachedNeedsConditions`

Zalathar committed:
Declare a `run-coverage-rustdoc` suite for coverage tests that need `rustdoc`

Zalathar committed:
Declare a `run-coverage` test mode/suite in bootstrap

Zalathar committed:
Migrate most of the existing coverage tests over to `run-coverage`

Zalathar committed:
Re-bless the newly-migrated tests

Zalathar committed:
Remove the old `coverage-reports` and `coverage` directories

Zalathar commented:
Looks like this is all working again now.

Zalathar commented:
Dealing with `doctest.rs` was tricky, since it’s inherently a lot more complicated than the other tests.

- I was tempted to leave it as its own run-make test, but that would ultimately leave us with make/shell/Python code duplicating much of run-coverage.
- I thought about including it in the main run-coverage suite, but that would force the whole suite to require rustdoc (which is a real cost from a clean build), and require a new directive to actually trigger running the doctests (or running them unconditionally for everything).

So while it feels a bit weird to introduce a whole extra suite for one test, I think it ends up being the more natural solution.

wesleywiser commented on src/tools/compiletest/src/header.rs:
```diff
@@ -894,7 +909,27 @@ pub fn make_test_description<R: Read>(
+        Mode::RunCoverage => {
+            &[
+                "needs-profiler-support",
+                // FIXME(mati865): MinGW GCC miscompiles compiler-rt profiling library but with Clang it works
+                // properly. Since we only have GCC on the CI ignore the test for now.
+                "ignore-windows-gnu",
+                // FIXME(pietroalbini): this test currently does not work on cross-compiled
+                // targets because remote-test is not capable of sending back the *.profraw
+                // files generated by the LLVM instrumentation.
+                "ignore-cross-compile",
```
Thanks for including these FIXMEs here, that will make it easier to understand in the future when they can be removed!

wesleywiser commented on tests/run-coverage/uses_crate.rs:
```diff
@@ -1,8 +1,12 @@
 // FIXME #110395
-// ignore-llvm-cov-show-diffs
+// ignore-linux
```
I'm not sure this has the same effect as the test will still run on some platforms like macOS and Windows. Should this just be `// ignore` for now?

Zalathar replied:
Ah yes, this deserves more explanation.

I’ve verified that these tests actually pass in CI on macOS and Windows. It seems that they are only broken on Linux, but the makefile didn’t have a way to ignore a single tests on a single platform, so when they broke on Linux they ended up being disabled on all platforms.

(To be honest, I changed this to `ignore-linux` because I didn’t know about plain `ignore`. I do think it’s fine to re-enable the tests on other platforms, but I’m willing to change it if you still think this is a bad idea.)

Zalathar replied:
For more context, these tests seem to have been broken/disabled by #110393, which is kind of baffling and I don’t know if anyone has had a chance to properly investigate what happened. I suspect it has something to do with optimisation.

wesleywiser replied:
Yeah, I was also confused to see that the ignore came from #110393. If these tests pass on `!linux`, then by all means, let's run them!

wesleywiser commented on tests/run-coverage/uses_inline_crate.rs:
```diff
@@ -1,10 +1,12 @@
 // FIXME #110395
-// ignore-llvm-cov-show-diffs
+// ignore-linux
```
Same comment here

wesleywiser reviewed (commented):
This is looking pretty good to me! I noticed there used to be a mechanism in the coverage makefile to bless the tests, does `--bless` still work?

Zalathar commented:
> This is looking pretty good to me! I noticed there used to be a mechanism in the coverage makefile to bless the tests, does `--bless` still work?

Yes, `--bless` is now handled automatically by the compiletest methods for comparing output. I’ve used it on these migrated tests to manually verify that it still works (e.g. in the patch that re-blesses some tests).

Zalathar commented on src/bootstrap/test.rs:
```diff
@@ -1299,6 +1299,13 @@ host_test!(RunMakeFullDeps {
 
 default_test!(Assembly { path: "tests/assembly", mode: "assembly", suite: "assembly" });
 
+host_test!(RunCoverage { path: "tests/run-coverage", mode: "run-coverage", suite: "run-coverage" });
+host_test!(RunCoverageRustdoc {
+    path: "tests/run-coverage-rustdoc",
+    mode: "run-coverage",
+    suite: "run-coverage-rustdoc"
+});
+
```
I'm not fully clear on the differences between `default_test!` and `host_test!`, but since the tests are skipped on cross-compile targets anyway I assume it doesn't matter much right now.

(It should be possible to run coverage tests on cross-compile targets, but the cross-runner would need to be taught how to send back the resulting `.profraw` files to be processed by the host.)

Zalathar replied:
I also don't have strong feelings on the name of the extra suite; I was wavering between `run-coverage-doctest`, `run-coverage-doctests`, and `run-coverage-rustdoc`.

(If there are ever other run-coverage tests with extra tool dependencies, it would probably be fine to group them all into one extra suite with a more generic name.)

Zalathar commented on src/bootstrap/test.rs:
```diff
@@ -1598,17 +1606,21 @@ note: if you're sure you want to do this, please open an issue as to why. In the
                     .expect("Expected llvm-config to be contained in directory");
                 assert!(llvm_bin_path.is_dir());
                 cmd.arg("--llvm-bin-dir").arg(llvm_bin_path);
+            }
 
+            if !builder.config.dry_run() && matches!(suite, "run-make" | "run-make-fulldeps") {
```
The new mode/suites don't need LLD, so I've split this off into a separate conditional that only checks for the run-make suites.

bors commented:
:umbrella: The latest upstream changes (presumably #112418) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
(Rebased to resolve a false merge conflict in bootstrap; no changes.)

Zalathar commented:
Out of curiosity I tried re-enabling `uses_crate.rs` and `uses_inline_crate.rs` on Linux (on a separate branch), and was surprised to find that they appear to succeed in CI on `x86-64_gnu`.

(They appear to succeed both with my changes, and also on master without my changes.)

I know I saw them fail in https://github.com/rust-lang/rust/pull/111179#issuecomment-1537270712 , so this is a bit mysterious.

jyn514 commented:
ok at this point i think it is clear i'm not going to get around to this PR in a reasonable amount of time.

r? bootstrap

Zalathar commented:
I'm going to keep rebasing this as needed to make my life easier while I work on other coverage-related stuff on top of it. But for ease of review I plan to leave the underlying patches unchanged.

bors commented:
:umbrella: The latest upstream changes (presumably #112817) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser commented:
@Zalathar this seems like it's ready to merge to me. Are there any loose ends I'm missing that should be resolved before I r+? 

Thanks for your work on this, it's really awesome to see these tests get converted away from using Makefiles!!

Zalathar commented:
@wesleywiser I consider this ready to merge. Other than rebasing, I don't plan to make any more changes unless reviewers ask for them.

There are no loose ends that would block merging, though I do want to highlight two things:

- It's possible to bikeshed things like the name of `run-coverage-rustdoc`, or the relative order of the new test suites in the list of tests. But I think I've made reasonable choices, and if anyone else has further input it should be easy enough to change things later.

- I've seen some indication that it *might* be possible to get rid of `// ignore-linux`, but I only have low confidence in that, and I'm not in a position to easily investigate whether/why the tests magically resumed working on Linux. So I'd prefer to get this merged as-is, with those tests now running on Windows/Mac instead of on no platforms, and then maybe open a tiny follow-up PR to re-enable them on Linux too.

So overall, this should be good to merge in its current state.

bors commented:
:umbrella: The latest upstream changes (presumably #113105) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser commented:
Ok, this all looks good to me! I agree, I don't think either of those points need to be resolved prior to merging this (the naming and ordering of the current patch seems totally fine and if someone thinks of a better alternative, they just send a PR that implements it). 

Thank you again for your work on this!
@bors r+

bors commented:
:pushpin: Commit 7b4e75b98965f45dd35a31968d388c7deca997a2 has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser 7b4e75b98965f45dd35a31968d388c7deca997a2 -->
<!-- homu: {"type":"Approved","sha":"7b4e75b98965f45dd35a31968d388c7deca997a2","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
